test/Compiler-filesystem-caching.test.js

```diff
@@ -0,0 +1,152 @@
+			const c = webpack(options);
+			c.hooks.compilation.tap(
+				"CompilerCachingTest",
+				compilation => (compilation.bail = true)
+			);
+			c.run((err, stats) => {
+				if (err) throw err;
+				expect(typeof stats).toBe("object");
+				stats = stats.toJson({
+					modules: true,
+					reasons: true
+				});
+				expect(typeof stats).toBe("object");
+				expect(stats).toHaveProperty("errors");
+				expect(Array.isArray(stats.errors)).toBe(true);
+				if (stats.errors.length > 0) {
+					onError(new Error(JSON.stringify(stats.errors, null, 4)));
+				}
+				c.close(() => {
+					onSuccess(stats);
+				});
+			});
+		}
+
+		runCompiler(onSuccess, onError);
+
+		return {
+			runAgain: runCompiler
+		};
+	}
+
+	function cleanup() {
+		rimraf.sync(`${tempFixturePath}*`);
+	}
+
+	beforeAll(cleanup);
+	afterAll(cleanup);
+
+	function createTempFixture() {
+		const fixturePath = `${tempFixturePath}-${fixtureCount}`;
+		const usesAssetFilepath = path.join(fixturePath, "uses-asset.js");
+		const svgFilepath = path.join(fixturePath, "file.svg");
+
+		// Remove previous copy if present
+		rimraf.sync(fixturePath);
+
+		// Copy over file since we"ll be modifying some of them
+		fs.mkdirSync(fixturePath);
+		fs.copyFileSync(
+			path.join(__dirname, "fixtures", "uses-asset.js"),
+			usesAssetFilepath
+		);
+		fs.copyFileSync(path.join(__dirname, "fixtures", "file.svg"), svgFilepath);
+
+		fixtureCount++;
+		return {
+			rootPath: fixturePath,
+			usesAssetFilepath: usesAssetFilepath,
+			svgFilepath: svgFilepath
+		};
+	}
+
+	it("should compile again when cached asset has changed but loader output remains the same", done => {
+		const tempFixture = createTempFixture();
+
+		const onError = error => done(error);
+
+		const helper = compile(
+			tempFixture.usesAssetFilepath,
+			stats => {
+				// Not cached the first time
+				expect(stats.assets[0].name).toBe("bundle.js");
+				expect(stats.assets[0].emitted).toBe(true);
+
+				expect(stats.assets[1].name).toMatch(/\w+\.svg$/);
+				expect(stats.assets[0].emitted).toBe(true);
+
+				helper.runAgain(stats => {
+					// Cached the second run
+					expect(stats.assets[0].name).toBe("bundle.js");
+					expect(stats.assets[0].emitted).toBe(false);
+
+					expect(stats.assets[1].name).toMatch(/\w+\.svg$/);
+					expect(stats.assets[0].emitted).toBe(false);
+
+					const svgContent = fs
+						.readFileSync(tempFixture.svgFilepath)
+						.toString()
+						.replace("icon-square-small", "icon-square-smaller");
+
+					fs.writeFileSync(tempFixture.svgFilepath, svgContent);
+
+					helper.runAgain(stats => {
+						// Still cached after file modification because loader always returns empty
+						expect(stats.assets[0].name).toBe("bundle.js");
+						expect(stats.assets[0].emitted).toBe(false);
+
+						expect(stats.assets[1].name).toMatch(/\w+\.svg$/);
+						expect(stats.assets[0].emitted).toBe(false);
+
+						done();
+					}, onError);
+				}, onError);
+			},
+			onError
+		);
+	});
+});
```

Comment:
https://github.com/webpack/webpack/tree/main/test/watchCases/cache/asset-modules, so you don't need to write extra lines and it will be easy to maintance tests in future, thank you

Reply:
I originally attempted to write a test under asset-modules but could not get it to work, as far as I can see there isn't a mechanism in that test suite to give me the ability to modify an asset between compile steps. I had to heavily tweak the compilation approach in this unit test such that it:
- Used a new `webpack()` instance each time, to avoid relying on shared in-memory state
- Used a physical filesystem, so that cache could be written out and re-used between builds
- Made a deliberate modification to a source file between compile passes

I don't think these three things are possible under `asset-modules`